src/dmd/typesem.d

```diff
@@ -3223,7 +3223,24 @@ private extern(C++) final class DotExpVisitor : Visitor
                 }
                 checkAccess(e.loc, sc, null, v);
                 Expression ve = new VarExp(e.loc, v);
-                ve = ve.expressionSemantic(sc);
+                // In case the expression is something known trivial, like this.enum, super.enum,
+                // variable.enum or type.enum, where we know the lhs does not have side effects,
+                // just use the enum directly.
+                // It would be preferable if we could just directly check if the left hand side of the
+                // expression is obviously side-effect free, but there does not seem to be a way to do that.
```

Comment:
See https://github.com/dlang/dmd/blob/master/src/dmd/sideeffect.d

Reply:
As I understand it, a function that never returns but is pure will still be considered as not having side effects by this code? But we must not optimize it away: we may optimize >1 calls to 1 call, but we mustn't optimize 1 call to 0 calls.